### PR TITLE
Parse data from TABLES/TABLE in DIAG-DATA-DICTIONARY-SPEC

### PR DESCRIPTION
**Pull request by smileprem**
- Parse the data from TABLES and persist them in id_lookup dict to be used elsewhere
- Added three utility functions to fetch the addresses for CAN and DoIP
- Updated .gitignore to ignore specific folders.

### CODE REVIEW

**Comment by smileprem** [comment]
image

Hi team - Could you please update the email address to send CLA?

**Comment by andlaus** [comment]
Hi Premkumar,

Thanks for your contribution and sorry for the sluggishness; I was AFK last week, so I missed this PR until now. I promise to assign the attention it deserves to it next week :)

You can try to send the CLA to cla-mbition@mercedes-benz.com or, if this does not work either, to me personally: andreas.lauser@mercedes-benz.com .

**Comment by smileprem** [comment]
Hi @andlaus - I sent the CLA to the email addresses mentioned above. Thanks for making this wonderful library as open source. I'll contribute back to the library as and when I can. Thank you.

File: odxtools/table.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+@dataclass()
+class TableRow:
+    """This class represents a TABLE-ROW."""
+
+    id: str
+    short_name: str
+    long_name: str
+    key: int
+    structure_ref: str
```
Comment (andlaus) [inline odxtools/table.py:29]: you should probably resolve references. the mechanism is to add a `_resolve_references(id_lookup)` method to all classes containing references and call it from the instancing class' `_respolve_references()`. in this case, you should add these method to `TableRow` and `Table` and call the latter one from  [DataDictionarySpec](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/diagdatadictionaryspec.py#L75).

**Review by andlaus — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution, we appreciate it :)

The parsing code is mostly fine, but before it is merged, I'd like to see support for writing table objects (see e.g. the writing code for the [audience data](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/pdx_stub/macros/printAudience.tpl)) and an extension of the ["Somersault" ECU](https://github.com/mercedes-benz/odxtools/blob/main/examples/somersaultecu.py). once this is done, extending [test_somersault](https://github.com/mercedes-benz/odxtools/blob/main/tests/test_somersault.py) would be nice.

File: CONTRIBUTORS_LICENSE_AGREEMENT.md [hunk 2]
```diff
@@ -34,6 +34,7 @@ Germany
 (_Contributor_ and _FOSS Co_ hereinafter individually a **_“Party”_** and jointly the **_“Parties”_**)
 
 ## 1 Definitions
+
```
[the remark this answers is not in the document]
Comment (smileprem) [reply CONTRIBUTORS_LICENSE_AGREEMENT.md:37]: Sure. I'll revert the changes.

File: README.md [hunk 3]
```diff
@@ -483,10 +483,15 @@ The included unit tests can be run via
 $ python -m unittest tests/test_*.py
 ```
 
+The static type checker can be run via
+```
```
[the remark this answers is not in the document]
Comment (smileprem) [reply README.md:487]: Updated

File: README.md [hunk 4]
```diff
@@ -483,10 +483,15 @@ The included unit tests can be run via
...
+$ python3 -m mypy --ignore-missing-imports --python-version 3.8 odxtools
```
[the remark this answers is not in the document]
Comment (smileprem) [reply README.md:488]: Updated

File: odxtools/parameters/readparameter.py [hunk 5]
```diff
@@ -193,10 +194,11 @@ def read_parameter_from_odx(et_element):
                                  table_snref=table_snref,
                                  table_row_snref=row_snref,
                                  table_row_ref=row_ref,
+                                 id=parameter_id,
                                  long_name=long_name,
-                                 semantic=semantic,
                                  byte_position=byte_position,
                                  bit_position=bit_position,
+                                 semantic=semantic,
```
[the remark this answers is not in the document]
Comment (smileprem) [reply odxtools/parameters/readparameter.py:201]: To reflect the same order as the class TableKeyParameter. Happy to revert if you think its not required.

```
class TableKeyParameter(Parameter):
    def __init__(self,
                 short_name,
                 table_ref=None,
                 table_snref=None,
                 table_row_snref=None,
                 table_row_ref=None,
                 id=None,
                 long_name=None,
                 byte_position=None,
                 bit_position=0,
                 semantic=None,
                 description=None):
```

File: odxtools/table.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2022 MBition GmbH
```
[the remark this answers is not in the document]
Comment (smileprem) [reply odxtools/table.py:2]: Thanks for the offer. But, I think it should be OK. Its a very small MR. :)

File: odxtools/table.py (hunk 1 above)
Comment (smileprem) [reply odxtools/table.py:29]: Sure, I'll add the recommended approach.

File: odxtools/parameters/readparameter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (andlaus) [reply odxtools/parameters/readparameter.py:201]: no, that's okay. I was just wondering...

File: odxtools/table.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andlaus) [reply odxtools/table.py:2]: well, you created that file :)

Anyway, If i underdstand the legalese correctly, these copyright notices at the beginning of source files are purely informal anyway (the copyright is with whomever made the respective piece of work...)

File: odxtools/table.py (hunk 1 above)
Comment (smileprem) [reply odxtools/table.py:30]: Hi @andlaus - I've added the functionality to resolve the references for TABLE and TABLE-ROWS. Please review. 

I'll review the templates and the test cases for tables in the next few days.

File: odxtools/table.py [hunk 7]
```diff
@@ -28,6 +29,21 @@ class TableRow:
...
+    def _resolve_references(self, id_lookup: Dict[str, Any]) -> None:
+        self._structure = id_lookup.get(self.structure_ref)
+        if self._structure is None:
+            logger.warning(
+                f"STRUCTURE-REF '{self.structure_ref!r}' could not be resolved."
+            )
+
+    @property
+    def structure(self) -> Optional[DopBase]:
+        """The data object property describing this parameter."""
```
Comment (andlaus) [inline odxtools/table.py:44]:
```suggestion
        """The structure associated with this table."""
```

(looks like a copy-and-pasto to me...)

File: odxtools/table.py [hunk 8]
```diff
@@ -51,12 +67,25 @@ def __init__(
             id=id, short_name=short_name, long_name=long_name, key_dop_ref=key_dop_ref
         )
         self.table_rows = table_rows
+        self._key_dop = None
+
+    @property
+    def key_dop(self) -> Optional[DopBase]:
+        """The data object property describing this parameter."""
```
Comment (andlaus) [inline odxtools/table.py:74]: this docstring is also wrong?

**Review by andlaus — COMMENTED** [review]
the new reference resolution logic looks good. Thanks for the work :)

File: odxtools/table.py (hunk 8 above)
Comment (smileprem) [reply odxtools/table.py]: Updated the doc string

**Comment by smileprem** [comment]
> Thanks a lot for your contribution, we appreciate it :)
> 
> The parsing code is mostly fine, but before it is merged, I'd like to see support for writing table objects (see e.g. the writing code for the [audience data](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/pdx_stub/macros/printAudience.tpl)) and an extension of the ["Somersault" ECU](https://github.com/mercedes-benz/odxtools/blob/main/examples/somersaultecu.py). once this is done, extending [test_somersault](https://github.com/mercedes-benz/odxtools/blob/main/tests/test_somersault.py) would be nice.

- Added the code to write tables to pdx files
- Updated the test case in test_diag_data_dictionary_spec.py to include tables

File: examples/somersaultecu.py [hunk 9]
```diff
@@ -364,6 +361,40 @@ class SID(IntEnum):
         compu_method=somersault_compumethods["boolean"]),
 }
 
+# tables
+somersault_tables = {
```
Comment (smileprem) [inline examples/somersaultecu.py:365]: Please feel free to change the content of this test table

**Comment by andlaus** [comment]
Thanks for the fixes. this looks very good now and from the coding point of view it is basically ready to be merged. There's one issue which prevents me from hitting green, though: the new generated PDX files do no longer validate:

```bash
$ ./mksomersaultpdx.py somersault.pdx
$ unzip somersault.pdx
$ xmllint --schema ~/standards/odx/BaseStandard/Schema/odx.xsd --noout somersault.odx-d
somersault.odx-d:179: element DATA-OBJECT-PROPS: Schemas validity error : Element 'DATA-OBJECT-PROPS': This element is not expected. Expected is ( SDGS ).
somersault.odx-d fails to validate
```

(This is not a big deal, most likely it is caused by a wrong order of the XML tags. if you don't have the XSD available, I can fix this for you.)

besides this, there are also two nice-to-haves before merge: it would be good if the shipped version of `somersault.pdx` would correspond to the current state of affairs (just run `mksomersault.py` in the examples directory and `git add` the result), and it would be great if the commits contained in this PR would be squashed into one (I can do this via a squash merge, but then the commit description would be of very limited use).

File: examples/somersaultecu.py (hunk 9 above)
Comment (andlaus) [reply examples/somersaultecu.py:365]: it is fine (quirky) :)

**Comment by smileprem** [comment]
Thanks @andlaus. 

- Updated the order in which tables were inserted, thus passing the validation
- Added `somersault.pdx` to the codebase
- Added a README file with the instructions to validate the pdx generation logic

We generally squash the commits during merge. Please feel free to do so. Also, it would be nice to add the odx.xsd to the codebase. We could eventually add schema validation in the parsing logic too. What do you think?

**Comment by andlaus** [comment]
Good job! Merging, thanks for your persistence!

**Pull request merged**